File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/java/com/tencent/devops/scm/provider/git/tgit/TGitWebhookEnricher.java
package com.tencent.devops.scm.provider.git.tgit;

import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_MR_SOURCE_BRANCH;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_MR_TARGET_BRANCH;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_OPERATION_KIND;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.CI_BRANCH;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_BASE_REF;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_AUTHOR;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_MESSAGE;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_HEAD_REF;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_MR_TAPD_ISSUES;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_REF;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA_SHORT;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_TAG_DESC;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_WEBHOOK_BRANCH;
import static com.tencent.devops.scm.sdk.tgit.enums.TGitPushOperationKind.UPDATE_NONFASTFORWORD;

import com.tencent.devops.scm.api.WebhookEnricher;
import com.tencent.devops.scm.api.function.PairConsumer;
import com.tencent.devops.scm.api.pojo.Change;
import com.tencent.devops.scm.api.pojo.PullRequest;
import com.tencent.devops.scm.api.pojo.Review;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository;
import com.tencent.devops.scm.api.pojo.repository.git.GitScmServerRepository;
import com.tencent.devops.scm.api.pojo.webhook.Webhook;
import com.tencent.devops.scm.api.pojo.webhook.git.AbstractCommentHook;
import com.tencent.devops.scm.api.pojo.webhook.git.CommitCommentHook;
import com.tencent.devops.scm.api.pojo.webhook.git.GitPushHook;
import com.tencent.devops.scm.api.pojo.webhook.git.GitTagHook;
import com.tencent.devops.scm.api.pojo.webhook.git.IssueCommentHook;
import com.tencent.devops.scm.api.pojo.webhook.git.IssueHook;
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestCommentHook;
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestHook;
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestReviewHook;
import com.tencent.devops.scm.sdk.tgit.TGitApiException;
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory;
import com.tencent.devops.scm.sdk.tgit.enums.TGitTapdWorkType;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitCommit;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitDiff;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequest;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitProject;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitReview;

import com.tencent.devops.scm.sdk.tgit.pojo.TGitTag;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitTapdWorkItem;
import com.tencent.devops.scm.sdk.tgit.util.TGitStringHelper;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TGitWebhookEnricher implements WebhookEnricher {

    private final TGitApiFactory apiFactory;
    private final Map<Class<? extends Webhook>, PairConsumer<GitScmProviderRepository, Webhook>> eventActions;

    private static final Logger logger = LoggerFactory.getLogger(TGitApiException.class);

    public TGitWebhookEnricher(TGitApiFactory apiFactory) {
        this.apiFactory = apiFactory;
        eventActions = new HashMap<>();
        eventActions.put(PullRequestHook.class, this::enrichPullRequestHook);
        eventActions.put(GitPushHook.class, this::enrichPushHook);
        eventActions.put(IssueHook.class, this::enrichIssueHook);
        eventActions.put(PullRequestReviewHook.class, this::enrichPullRequestViewHook);
        eventActions.put(CommitCommentHook.class, this::enrichNoteHook);
        eventActions.put(IssueCommentHook.class, this::enrichNoteHook);
        eventActions.put(PullRequestCommentHook.class, this::enrichNoteHook);
        eventActions.put(GitTagHook.class, this::enrichTagHook);
    }

    @Override
    public Webhook enrich(ScmProviderRepository repository, Webhook webhook) {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        enrichServerRepository(repo, webhook);
        if (eventActions.containsKey(webhook.getClass())) {
            eventActions.get(webhook.getClass()).accept(repo, webhook);
        }
        return webhook;
    }

    private void enrichServerRepository(GitScmProviderRepository repo, Webhook webhook) {
        TGitApiTemplate.execute(repo, apiFactory, (providerRepository, tGitApi) -> {
            TGitProject project = tGitApi.getProjectApi().getProject(repo.getProjectIdOrPath());
            GitScmServerRepository serverRepository = (GitScmServerRepository) webhook.repository();
            serverRepository.setDefaultBranch(project.getDefaultBranch());
            serverRepository.setArchived(project.getArchived());
        });
    }

    private void enrichPullRequestHook(GitScmProviderRepository repository, Webhook webhook) {
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
                    PullRequestHook pullRequestHook = (PullRequestHook) webhook;
                    PullRequest pullRequest = pullRequestHook.getPullRequest();
                    TGitMergeRequest mergeRequest = tGitApi.getMergeRequestApi()
                            .getMergeRequestChanges(repository.getProjectIdOrPath(), pullRequest.getId());
                    if (mergeRequest != null) {
                        List<Change> changes = mergeRequest.getFiles()
                                .stream()
                                .map(TGitObjectConverter::convertChange)
                                .collect(Collectors.toList());
                        pullRequestHook.setChanges(changes);
                    }
                    // 根据完整的MR/Review信息填充变量
                    pullRequestHook.getExtras().putAll(
                            fillPullRequestReviewVars(
                                    repository,
                                    pullRequest,
                                    ((PullRequestHook) webhook).getRepo()
                            )
                    );
                }
        );
    }

    private void enrichPushHook(GitScmProviderRepository repository, Webhook webhook) {
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            GitPushHook pushHook = (GitPushHook) webhook;
            String operationKind = (String) pushHook.getExtras().get(BK_REPO_GIT_WEBHOOK_PUSH_OPERATION_KIND);
            if (UPDATE_NONFASTFORWORD.value.equals(operationKind)) {
                List<TGitDiff> diffs = tGitApi.getRepositoryFileApi()
                        .getFileChange(
                                repository.getProjectIdOrPath(),
                                pushHook.getAfter(),
                                pushHook.getBefore(),
                                false,
                                false,
                                1,
                                1000
                        );
                if (CollectionUtils.isEmpty(diffs)) {
                    return;
                }
                pushHook.setChanges(diffs.stream()
                        .map(TGitObjectConverter::convertChange)
                        .collect(Collectors.toList()));
            }
        });
    }

    private void enrichIssueHook(GitScmProviderRepository repository, Webhook webhook) {
        IssueHook issueHook = (IssueHook) webhook;
        if (issueHook.getExtras() == null) {
            issueHook.setExtras(new HashMap<>());
        }
        issueHook.getExtras().putAll(fillDefaultBranchVars(repository, webhook, null));
    }

    private void enrichNoteHook(GitScmProviderRepository repository, Webhook webhook) {
        AbstractCommentHook commentHook = (AbstractCommentHook) webhook;
        if (commentHook.getExtras() == null) {
            commentHook.setExtras(new HashMap<>());
        }
        // 默认分支的其他信息
        Map<String, Object> extras = commentHook.getExtras();
        Map<String, String> defaultBranchVars = fillDefaultBranchVars(
                repository,
                webhook,
                (defaultBranch, commit) -> {
                    if (defaultBranch != null && commit != null) {
                        extras.put(
                                PIPELINE_GIT_COMMIT_MESSAGE,
                                StringUtils.defaultString(commit.getMessage(), "")
                        );
                    }
                }
        );
        extras.putAll(defaultBranchVars);
        if (commentHook instanceof PullRequestCommentHook
                && ((PullRequestCommentHook) commentHook).getPullRequest() != null) {
            // 仅与MR关联的Review事件才需调用API
            PullRequest pullRequest = ((PullRequestCommentHook) commentHook).getPullRequest();
            extras.putAll(
                    fillPullRequestReviewVars(
                            repository,
                            pullRequest,
                            ((PullRequestCommentHook) webhook).getRepo()
                    )
            );
        }
        commentHook.setExtras(extras);
    }

    private void enrichTagHook(GitScmProviderRepository repository, Webhook webhook) {
        GitTagHook tagHook = (GitTagHook) webhook;
        tagHook.getExtras().putAll(
                fillTagVars(
                        repository,
                        tagHook.getRef().getName()
                )
        );
    }

    private void enrichPullRequestViewHook(
            GitScmProviderRepository repository,
            Webhook webhook
    ) {
        PullRequestReviewHook reviewHook = (PullRequestReviewHook) webhook;
        PullRequest pullRequest = reviewHook.getPullRequest();
        Map<String, Object> extras = reviewHook.getExtras();
        if (pullRequest != null) {
            extras.putAll(
                    fillPullRequestReviewVars(
                            repository,
                            pullRequest,
                            ((PullRequestReviewHook) webhook).getRepo()
                    )
            );
        } else {
            TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
                Review review = reviewHook.getReview();
                TGitReview commitReview = tGitApi.getReviewApi()
                        .getCommitReview(
                                repository.getProjectIdOrPath(),
                                review.getId()
                        );
                review.setTitle(StringUtils.defaultString(commitReview.getTitle(), ""));
            });
        }
        // 默认分支信息
        Map<String, String> defaultBranchVars = fillDefaultBranchVars(repository, webhook,
                (defaultBranch, commit) -> {
                    if (defaultBranch != null && commit != null) {
                        extras.put(
                                PIPELINE_GIT_COMMIT_MESSAGE,
                                StringUtils.defaultString(commit.getMessage(), "")
                        );
                    }
                }
        );
        extras.putAll(defaultBranchVars);
        // 目前BASE_REF和HEAD_REF分别代表目标分支和源分支
        extras.put(PIPELINE_GIT_BASE_REF, extras.get(BK_REPO_GIT_WEBHOOK_MR_TARGET_BRANCH));
        extras.put(PIPELINE_GIT_HEAD_REF, extras.get(BK_REPO_GIT_WEBHOOK_MR_SOURCE_BRANCH));
        extras.putAll(
                fillTapdWorkItemsVar(
                        repository,
                        TGitTapdWorkType.MR,
                        ((PullRequestReviewHook) webhook).getPullRequest().getNumber().longValue()
                )
        );
    }

    /**
     * 填充MR变量信息
     * @param pullRequest pull request 基础信息，webhook body中获取的基础信息
     */
    private Map<String, Object> fillPullRequestVars(
            GitScmProviderRepository repository,
            PullRequest pullRequest,
            GitScmServerRepository scmServerRepository
    ) {
        Map<String, Object> outputs = new HashMap<>();
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            TGitMergeRequest mergeRequestInfo = tGitApi.getMergeRequestApi()
                    .getMergeRequestById(
                            repository.getProjectIdOrPath(),
                            pullRequest.getId()
                    );
            if (mergeRequestInfo != null) {
                pullRequest.setTitle(StringUtils.defaultString(mergeRequestInfo.getTitle(), ""));
                outputs.putAll(TGitObjectToMapConverter.convertPullRequest(mergeRequestInfo, scmServerRepository));
            }
        });
        return outputs;
    }

    /**
     * 填充与Mr关联的Review参数
     */
    private Map<String, Object> fillPullRequestReviewVars(
            GitScmProviderRepository repository,
            PullRequest pullRequest,
            GitScmServerRepository scmServerRepository
    ) {
        Map<String, Object> vars = new HashMap<>(
                fillPullRequestVars(
                        repository,
                        pullRequest,
                        scmServerRepository
                )
        );

        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            TGitReview tGitMergeRequestReviewInfo = tGitApi.getReviewApi()
                    .getMergeRequestReview(
                            repository.getProjectIdOrPath(),
                            pullRequest.getId()
                    );
            if (tGitMergeRequestReviewInfo != null) {
                vars.putAll(TGitObjectToMapConverter.convertReview(tGitMergeRequestReviewInfo));
            }
        });
        return vars;
    }

    /**
     * 获取默认分支变量
     * @param repository 仓库信息
     * @param webhook webhook信息
     * @param extAction 扩展action, 用于扩展默认分支和commit信息
     * @return 默认分支和commit信息（基础变量，不包含扩展字段，扩展字段需在外部手动组装）
     */
    private Map<String, String> fillDefaultBranchVars(
            GitScmProviderRepository repository,
            Webhook webhook,
            PairConsumer<String, TGitCommit> extAction
    ) {
        Map<String, String> extra = new HashMap<>();
        enrichDefaultBranchAndCommitInfo(
                repository,
                webhook,
                (defaultBranch, commit) -> {
                    if (defaultBranch != null && commit != null) {
                        extra.put(PIPELINE_GIT_REF, defaultBranch);
                        extra.put(CI_BRANCH, defaultBranch);
                        extra.put(PIPELINE_WEBHOOK_BRANCH, defaultBranch);
                        extra.put(PIPELINE_GIT_COMMIT_AUTHOR, commit.getAuthorName());
                        extra.put(PIPELINE_GIT_SHA, commit.getId());
                        extra.put(PIPELINE_GIT_SHA_SHORT, commit.getShortId());
                        // 额外字段，不同事件默认分支参数不同有差异，例如：Note 和 Issue
                        if (extAction != null) {
                            extAction.accept(defaultBranch, commit);
                        }
                    }
                }
        );
        return extra;
    }

    private void enrichDefaultBranchAndCommitInfo(
            GitScmProviderRepository repository,
            Webhook webhook,
            PairConsumer<String, TGitCommit> action
    ) {
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            GitScmServerRepository serverRepository = (GitScmServerRepository) webhook.repository();
            String defaultBranch = serverRepository.getDefaultBranch();
            TGitCommit commit = null;
            if (defaultBranch != null) {
                commit = tGitApi.getCommitsApi().getCommit(repository.getProjectIdOrPath(), defaultBranch);
            }
            action.accept(defaultBranch, commit);
        });
    }

    private Map<String, String> fillTapdWorkItemsVar(
            GitScmProviderRepository repository,
            TGitTapdWorkType type,
            Long iid
    ) {
        Map<String, String> extra = new HashMap<>();
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            List<TGitTapdWorkItem> tapdWorkItems = tGitApi.getProjectApi()
                    .getTapdWrorkItems(
                            repository.getProjectIdOrPath(),
                            type,
                            iid,
                            1,
                            100
                    );
            String tapdIds = TGitStringHelper.join(
                    CollectionUtils.emptyIfNull(tapdWorkItems)
                            .stream()
                            .map(TGitTapdWorkItem::getTapdId)
                            .collect(Collectors.toList())
            );
            if (type == TGitTapdWorkType.MR) {
                extra.put(PIPELINE_GIT_MR_TAPD_ISSUES, tapdIds);
            }
        });
        return extra;
    }

    private Map<String, String> fillTagVars(
            GitScmProviderRepository repository,
            String tagName
    ) {
        Map<String, String> extra = new HashMap<>();
        TGitApiTemplate.execute(repository, apiFactory, (providerRepository, tGitApi) -> {
            TGitTag tag = tGitApi.getTagsApi()
                    .getTag(
                            repository.getProjectIdOrPath(),
                            tagName
                    );
            if (tag != null) {
                extra.put(PIPELINE_GIT_TAG_DESC, tag.getDescription());
            }
        });
        return extra;
    }
}

File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/kotlin/com/tencent/devops/scm/provider/git/tgit/TGitWebhookEnricher.kt
package com.tencent.devops.scm.provider.git.tgit

import com.tencent.devops.scm.api.WebhookEnricher
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_MR_SOURCE_BRANCH
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_MR_TARGET_BRANCH
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_OPERATION_KIND
import com.tencent.devops.scm.api.constant.WebhookOutputCode.CI_BRANCH
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_BASE_REF
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_AUTHOR
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_MESSAGE
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_HEAD_REF
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_MR_TAPD_ISSUES
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_REF
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA_SHORT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_TAG_DESC
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_WEBHOOK_BRANCH
import com.tencent.devops.scm.api.pojo.PullRequest
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository
import com.tencent.devops.scm.api.pojo.repository.git.GitScmServerRepository
import com.tencent.devops.scm.api.pojo.webhook.Webhook
import com.tencent.devops.scm.api.pojo.webhook.git.AbstractCommentHook
import com.tencent.devops.scm.api.pojo.webhook.git.CommitCommentHook
import com.tencent.devops.scm.api.pojo.webhook.git.GitPushHook
import com.tencent.devops.scm.api.pojo.webhook.git.GitTagHook
import com.tencent.devops.scm.api.pojo.webhook.git.IssueCommentHook
import com.tencent.devops.scm.api.pojo.webhook.git.IssueHook
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestCommentHook
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestHook
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestReviewHook
import com.tencent.devops.scm.sdk.tgit.TGitApiException
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory
import com.tencent.devops.scm.sdk.tgit.enums.TGitPushOperationKind.UPDATE_NONFASTFORWORD
import com.tencent.devops.scm.sdk.tgit.enums.TGitTapdWorkType
import com.tencent.devops.scm.sdk.tgit.pojo.TGitCommit
import org.slf4j.LoggerFactory

@SuppressWarnings("TooManyFunctions")
class TGitWebhookEnricher(private val apiFactory: TGitApiFactory) : WebhookEnricher {

    private val eventActions: MutableMap<Class<out Webhook>, (GitScmProviderRepository, Webhook) -> Unit> =
        mutableMapOf()

    companion object {
        private val logger = LoggerFactory.getLogger(TGitApiException::class.java)
    }

    init {
        eventActions[PullRequestHook::class.java] = this::enrichPullRequestHook
        eventActions[GitPushHook::class.java] = this::enrichPushHook
        eventActions[IssueHook::class.java] = this::enrichIssueHook
        eventActions[PullRequestReviewHook::class.java] = this::enrichPullRequestViewHook
        eventActions[CommitCommentHook::class.java] = this::enrichNoteHook
        eventActions[IssueCommentHook::class.java] = this::enrichNoteHook
        eventActions[PullRequestCommentHook::class.java] = this::enrichNoteHook
        eventActions[GitTagHook::class.java] = this::enrichTagHook
    }

    override fun enrich(repository: ScmProviderRepository, webhook: Webhook): Webhook {
        val repo = repository as GitScmProviderRepository
        enrichServerRepository(repo, webhook)
        eventActions[webhook.javaClass]?.invoke(repo, webhook)
        return webhook
    }

    private fun enrichServerRepository(repo: GitScmProviderRepository, webhook: Webhook) {
        TGitApiTemplate.execute(repo, apiFactory) { _, tGitApi ->
            val project = tGitApi.projectApi.getProject(repo.projectIdOrPath)
            val serverRepository = webhook.repository() as GitScmServerRepository
            serverRepository.defaultBranch = project.defaultBranch
            serverRepository.archived = project.archived
        }
    }

    private fun enrichPullRequestHook(repository: GitScmProviderRepository, webhook: Webhook) {
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val pullRequestHook = webhook as PullRequestHook
            val pullRequest = pullRequestHook.pullRequest
            val mergeRequest = tGitApi.mergeRequestApi
                    .getMergeRequestChanges(
                        repository.projectIdOrPath,
                        pullRequest.id
                    )

            mergeRequest?.let {
                pullRequestHook.changes = it.files.map(TGitObjectConverter::convertChange)
            }
            // 根据完整的MR/Review信息填充变量
            pullRequestHook.extras.putAll(
                fillPullRequestReviewVars(
                    repository,
                    pullRequest,
                    webhook.repo
                )
            )
        }
    }

    private fun enrichPushHook(repository: GitScmProviderRepository, webhook: Webhook) {
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val pushHook = webhook as GitPushHook
            val operationKind = pushHook.extras[BK_REPO_GIT_WEBHOOK_PUSH_OPERATION_KIND] as String?

            if (UPDATE_NONFASTFORWORD.value == operationKind) {
                pushHook.changes = (tGitApi.repositoryFileApi.getFileChange(
                    repository.projectIdOrPath,
                    pushHook.after,
                    pushHook.before,
                    false,
                    false,
                    1,
                    1000
                ) ?: listOf()).map(TGitObjectConverter::convertChange)
            }
        }
    }

    private fun enrichIssueHook(repository: GitScmProviderRepository, webhook: Webhook) {
        val issueHook = webhook as IssueHook
        issueHook.extras.putAll(
            fillDefaultBranchVars(repository, webhook, null)
        )
    }

    private fun enrichNoteHook(repository: GitScmProviderRepository, webhook: Webhook) {
        val commentHook = webhook as AbstractCommentHook

        val extras = commentHook.extras
        val defaultBranchVars = fillDefaultBranchVars(
            repository,
            webhook
        ) { _, commit ->
            extras[PIPELINE_GIT_COMMIT_MESSAGE] = commit.message ?: ""
        }

        extras.putAll(defaultBranchVars)

        if (commentHook is PullRequestCommentHook) {
            commentHook.pullRequest?.let {
                extras.putAll(
                    fillPullRequestReviewVars(
                        repository,
                        it,
                        (webhook as PullRequestCommentHook).repo
                    )
                )
            }

        }
        commentHook.extras = extras
    }

    private fun enrichTagHook(repository: GitScmProviderRepository, webhook: Webhook) {
        val tagHook = webhook as GitTagHook
        tagHook.extras.putAll(
            fillTagVars(
                repository,
                tagHook.ref.name
            )
        )
    }

    private fun enrichPullRequestViewHook(repository: GitScmProviderRepository, webhook: Webhook) {
        val reviewHook = webhook as PullRequestReviewHook
        val pullRequest = reviewHook.pullRequest
        val extras = reviewHook.extras

        if (pullRequest != null) {
            extras.putAll(
                fillPullRequestReviewVars(
                    repository,
                    pullRequest,
                    webhook.repo
                )
            )
        } else {
            TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
                val review = reviewHook.review
                val commitReview = tGitApi.reviewApi.getCommitReview(
                    repository.projectIdOrPath,
                    review.id
                )
                review.title = commitReview.title ?: ""
            }
        }

        val defaultBranchVars = fillDefaultBranchVars(repository, webhook) { _, commit ->
            extras[PIPELINE_GIT_COMMIT_MESSAGE] = commit.message ?: ""
        }

        extras.putAll(defaultBranchVars)
    }

    private fun fillPullRequestVars(
        repository: GitScmProviderRepository,
        pullRequest: PullRequest,
        scmServerRepository: GitScmServerRepository
    ): Map<String, Any> {
        val outputs = mutableMapOf<String, Any>()
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val mergeRequestInfo = tGitApi.mergeRequestApi.getMergeRequestById(
                repository.projectIdOrPath,
                pullRequest.id
            )

            mergeRequestInfo?.let {
                pullRequest.title = it.title ?: ""
                outputs.putAll(TGitObjectToMapConverter.convertPullRequest(it, scmServerRepository))
            }
        }
        return outputs
    }

    private fun fillPullRequestReviewVars(
        repository: GitScmProviderRepository,
        pullRequest: PullRequest,
        scmServerRepository: GitScmServerRepository
    ): Map<String, Any> {
        val vars = mutableMapOf<String, Any>().apply {
            putAll(
                fillPullRequestVars(
                    repository,
                    pullRequest,
                    scmServerRepository
                )
            )
        }

        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val tGitMergeRequestReviewInfo = tGitApi.reviewApi.getMergeRequestReview(
                repository.projectIdOrPath,
                pullRequest.id
            )

            tGitMergeRequestReviewInfo?.let {
                vars.putAll(TGitObjectToMapConverter.convertReview(it))
            }
        }
        return vars
    }

    private fun fillDefaultBranchVars(
        repository: GitScmProviderRepository,
        webhook: Webhook,
        extAction: ((String, TGitCommit) -> Unit)?
    ): Map<String, String> {
        val extra = mutableMapOf<String, String>()
        enrichDefaultBranchAndCommitInfo(
            repository,
            webhook
        ) { defaultBranch, commit ->
            if (defaultBranch != null && commit != null) {
                extra[PIPELINE_GIT_REF] = defaultBranch
                extra[CI_BRANCH] = defaultBranch
                extra[PIPELINE_WEBHOOK_BRANCH] = defaultBranch
                extra[PIPELINE_GIT_COMMIT_AUTHOR] = commit.authorName
                extra[PIPELINE_GIT_SHA] = commit.id
                extra[PIPELINE_GIT_SHA_SHORT] = commit.shortId
                // 额外字段，不同事件默认分支参数不同有差异，例如：Note 和 Issue
                extAction?.invoke(defaultBranch, commit)
            }
        }
        return extra
    }

    private fun enrichDefaultBranchAndCommitInfo(
        repository: GitScmProviderRepository,
        webhook: Webhook,
        action: (String?, TGitCommit?) -> Unit
    ) {
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val serverRepository = webhook.repository() as GitScmServerRepository
            val defaultBranch = serverRepository.defaultBranch
            var commit: TGitCommit? = null
            if (defaultBranch != null) {
                commit = tGitApi.commitsApi.getCommit(repository.projectIdOrPath, defaultBranch)
            }
            action.invoke(defaultBranch, commit)
        }
    }

    private fun fillTapdWorkItemsVar(
        repository: GitScmProviderRepository,
        type: TGitTapdWorkType,
        iid: Long
    ): Map<String, String> {
        val extra = mutableMapOf<String, String>()
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val tapdWorkItems = tGitApi.projectApi.getTapdWrorkItems(
                repository.projectIdOrPath,
                type,
                iid,
                1,
                100
            )
            val tapdIds = tapdWorkItems?.joinToString(separator = ",") { it.tapdId.toString() }

            if (!tapdIds.isNullOrBlank() && type == TGitTapdWorkType.MR) {
                extra[PIPELINE_GIT_MR_TAPD_ISSUES] = tapdIds
            }
        }
        return extra
    }

    private fun fillTagVars(repository: GitScmProviderRepository, tagName: String): Map<String, String> {
        val extra = mutableMapOf<String, String>()
        TGitApiTemplate.execute(repository, apiFactory) { _, tGitApi ->
            val tag = tGitApi.tagsApi.getTag(
                repository.projectIdOrPath,
                tagName
            )
            tag?.let {
                extra[PIPELINE_GIT_TAG_DESC] = it.description
            }
        }
        return extra
    }
}
